components/ui/tabcounter/src/main/java/org/mozilla/components/ui/tabcounter/TabCounter.kt

```diff
@@ -0,0 +1,368 @@
+        val firstAnimator = animatorSet.childAnimations[0]
+
+        // Fadeout in 100ms, with firstAnimator (49~51, 2 frames).
+        val fadeOut = ObjectAnimator.ofFloat(text, "alpha",
+                ANIM_TEXT_FADEOUT_FROM, ANIM_TEXT_FADEOUT_TO).setDuration(ANIM_TEXT_FADEOUT_DURATION)
+
+        // FadeIn in 66 ms, after fadeOut with delay 96ms (57~61, 4 frames).
+        val fadeIn = ObjectAnimator.ofFloat(text, "alpha",
+                ANIM_TEXT_FADEIN_FROM, ANIM_TEXT_FADEIN_TO).setDuration(ANIM_TEXT_FADEIN_DURATION)
+        fadeIn.startDelay = (ANIM_TEXT_FADEIN_DELAY).toLong() // delay 6 frames after fadeOut
+
+        // Move down on y-axis, from 0 to 4.4 in 66ms, with fadeIn (57~61, 4 frames).
+        val moveDown = ObjectAnimator.ofFloat(text, "translationY",
+                ANIM_TEXT_MOVEDOWN_FROM, ANIM_TEXT_MOVEDOWN_TO).setDuration(ANIM_TEXT_MOVEDOWN_DURATION)
+        moveDown.startDelay = (ANIM_TEXT_MOVEDOWN_DELAY) // delay 6 frames after fadeOut
+
+        // Move up on y-axis, from 0 to 4.4 in 66ms, after moveDown (61~69, 8 frames).
+        val moveUp = ObjectAnimator.ofFloat(text, "translationY",
+                ANIM_TEXT_MOVEUP_FROM, ANIM_TEXT_MOVEUP_TO).setDuration(ANIM_TEXT_MOVEUP_DURATION)
+
+        animatorSet.play(firstAnimator).with(fadeOut)
+        animatorSet.play(fadeOut).before(fadeIn)
+        animatorSet.play(fadeIn).with(moveDown)
+        animatorSet.play(moveDown).before(moveUp)
+    }
+
+    private fun formatForDisplay(count: Int): String {
+        return if (count > MAX_VISIBLE_TABS) {
+            SO_MANY_TABS_OPEN
+        } else NumberFormat.getInstance().format(count.toLong())
+    }
+
+    private fun adjustTextSize(newCount: Int) {
+        val newRatio = if (newCount in TWO_DIGITS_TAB_COUNT_THRESHOLD..MAX_VISIBLE_TABS) {
+            TWO_DIGITS_SIZE_RATIO
+        } else {
+            ONE_DIGIT_SIZE_RATIO
+        }
+
+        if (newRatio != currentTextRatio) {
+            currentTextRatio = newRatio
+            text.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
+                override fun onGlobalLayout() {
+                    text.viewTreeObserver.removeOnGlobalLayoutListener(this)
+                    val sizeInPixel = (box.width * newRatio).toInt()
+                    if (sizeInPixel > 0) {
+                        // Only apply the size when we calculate a valid value.
+                        text.setTextSize(TypedValue.COMPLEX_UNIT_PX, sizeInPixel.toFloat())
+                    }
+                }
+            })
+        }
+    }
+
+    companion object {
+
+        internal const val MAX_VISIBLE_TABS = 99
+
+        internal const val SO_MANY_TABS_OPEN = "∞"
+        internal const val DEFAULT_TABS_COUNTER_TEXT = ":)"
+
+        internal const val ONE_DIGIT_SIZE_RATIO = 0.6f
+        internal const val TWO_DIGITS_SIZE_RATIO = 0.5f
+        internal const val TWO_DIGITS_TAB_COUNT_THRESHOLD = 10
+
+
```

Comment:
\[ktlint\] WARNING: \[no\-consecutive\-blank\-lines\] Needless blank line\(s\) in column 1